Automattic/wp-calypso
# Updates the navigation in the promote-post section

Related to SELFSERVE-576

We have a problem in the navigation when you switch sites while in the Advertising section. 

The current behavior is as follows:
- If you switch sites while you are on `/advertising/{SITE}` it will work ok
- If you click on a tab (posts, campaigns, credits) and then switch sites, it will first redirect you to your primary site landing page, and if you click it again it will switch site correctly.
- If you switch sites from the campaign details page it will load an incorrect page and give you an error. (The page that loads is `/advertising/{SITE}/campaigns/{NEW_SITE}`. 

This happened because the default behavior for navigating through sites in Calypso is by having the site slug at the end of the URL. Right now, we are assuming the site will be in the middle, and that giving us some problems.

## Proposed Changes

This PR changes how we handle navigation in the Advertising section by moving the site slug at the end of the route.

* Changes the controllers to have the `:site` param at the end (for Calypso and the Blaze Dashboard App)
* Changes all the links on the section to reference the new place of the site slug
* Added some compatibility routes to redirect the users if they got to the section using the previous navigation

Note: This is a significant change; let me know if you find a more straightforward way to fix this problem. I added compatibility redirects because we have some apps that are using the old navigation style. I will update those apps after we merge this PR. (The apps are the Jetpack Blaze module and the Create campaign widget).

## Testing Instructions

### Testing steps for Calypso

* Open the live preview and navigate to the Tools->Advertising section
* Switch sites and verify that the switch works correctly
* Click on a navigation tab, and then switch sites again. 
* Verify that the switch worked and that you now see the new site for that tab
* Go to the details for a campaign, and then switch sites
* Verify that you got redirected to the campaign list page for the new site
* Click on the promote button and then create a campaign
* Verify that you get redirected to the campaign list page after completing the campaign
* Enable the automatically redirect checkbox at the end step of the create campaign flow.
* Create another campaign and check that you got automatically redirected correctly to the campaign list page

### Testing steps for Jetpack Blaze Dashboard
You will need a working Jetpack installation. Then, in a console, execute this command on the `wp-calypso/apps/blaze-dashboard` folder:
`BLAZE_DASHBOARD_PACKAGE_PATH=/path_to_local_installation/jetpack/projects/packages/blaze yarn dev`
After that, you will be able to see this version of the app inside your local jetpack.


For Jetpack we need to verify that the navigation works correctly.
* Click on the Tools->Advertising page, and verify that the posts page shows correctly
* Click on the promote button in the header and on a post, and check that the widget loads correctly
* Then click on the Campaigns and credits tabs, and verify that those pages loads correctly too
* After that, go to the campaigns tab, and click on a campaign, and check that the details page works too.
* After this, we need to check that the redirection from other modules also work. For this we go to the Site's posts page by Clicking the menu, and then Posts (Menu->Posts->All posts)
* Click on the Blaze button for one of your posts
* Verify that you get redirected correctly to the create campaign widget for the selected post

## Pre-merge Checklist
